Assertion error when reshape follows Input layer in HGQ models

## Prerequisites
Please make sure to check off these prerequisites before submitting a bug report.
- [x] Test that the bug appears on the current version of the master branch. Make sure to include the commit hash of the commit you checked out.
- [x] Check that the issue hasn't already been reported, by checking the currently open issues.
- [x] If there are steps to reproduce the problem, make sure to write them down below.
- [x] If relevant, please include the hls4ml project files, which were created directly before and/or after the bug.

## Quick summary
The convert_from_keras_model function fails with an AssertionError when converting a Highly Granular Quantization (HGQ) Keras model that has a Reshape layer immediately following the Input layer. 

## Details

The bug is triggered by a specific post-training model state, not just by the architecture itself. While this state can be reached intermittently when training from scratch, the attached saved model provides a 100% reliable way to reproduce the failure.

The full error message is: 

`        
    hls_model = hls4ml.converters.convert_from_keras_model(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "./hls4ml_pr/hls4ml/hls4ml/utils/dependency.py", line 46, in inner
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "./hls4ml_pr/hls4ml/hls4ml/converters/__init__.py", line 223, in convert_from_keras_model
    return keras_v3_to_hls(config)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "./hls4ml_pr/hls4ml/hls4ml/converters/keras_v3_to_hls.py", line 295, in keras_v3_to_hls
    return ModelGraph.from_layer_list(config, layer_list, input_layers, output_layers)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "./hls4ml_pr/hls4ml/hls4ml/model/graph.py", line 451, in from_layer_list
    model.apply_flow(flow)
  File "./hls4ml_pr/hls4ml/hls4ml/model/graph.py", line 519, in apply_flow
    self._apply_sub_flow(flow, applied_flows)
  File "./hls4ml_pr/hls4ml/hls4ml/model/graph.py", line 528, in _apply_sub_flow
    self._apply_sub_flow(sub_flow, applied_flows)
  File "./hls4ml_pr/hls4ml/hls4ml/model/graph.py", line 528, in _apply_sub_flow
    self._apply_sub_flow(sub_flow, applied_flows)
  File "./hls4ml_pr/hls4ml/hls4ml/model/graph.py", line 531, in _apply_sub_flow
    applied_passes = optimize_model(self, flow.optimizers)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "./hls4ml_pr/hls4ml/hls4ml/model/optimizer/optimizer.py", line 319, in optimize_model
    res = opt.transform(model, node)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "./hls4ml_pr/hls4ml/hls4ml/model/optimizer/passes/bit_exact.py", line 838, in transform
    assert isinstance(
           ^^^^^^^^^^^
AssertionError: Input inputs_ connected to non-quantizer reshape_layer with non-trivial configuration

`

### Steps to Reproduce
Add what needs to be done to reproduce the bug. Add *commented* code examples and make sure to include the original model files / code, and the commit hash you are working on.
 
1. Setup the environment:
The following commands will create a conda environment, install the necessary dependencies, and clone the `hls4ml` repository at the specific commit on the main branch where the bug is observed.
```
mkdir -p ./hls4ml_pr
conda create --prefix ./hls4ml_pr/hls4ml_pr_env python=3.11
conda activate ./hls4ml_pr/hls4ml_pr_env
cd ./hls4ml_pr/
pip install HGQ2
git clone https://github.com/fastmachinelearning/hls4ml.git
cd hls4ml
git checkout 8a4f268
pip install -e .
python3 -m pip install 'tensorflow[and-cuda]'
pip install matplotlib scipy scikit-learn scikit-image pandas awkward uproot mplhep huggingface_hub calmjs tabulate pydot graphviz sympy
cd ..
```
2. Download and unzip the attached model file: `test.zip` 
[test.zip](https://github.com/user-attachments/files/21778059/test.zip)
3. cd into test folder and run using `python minimal_reshape.py`
### Expected behavior
Should print "Got the hls model successfully"

### Actual behavior
The error message pasted in the Details section above appears.

## Optional

### Possible fix
I have a fix prepared in a branch named `fix_reshape_issue` in my fork and will be opening a pull request to address this issue shortly.

### Additional context
We found this bug while working on a GSoC project. Mentor: Lino Gerlach @ligerlac

--- a/hls4ml/converters/keras_v3/merge.py
+++ b/hls4ml/converters/keras_v3/merge.py
@@ -39,7 +39,8 @@ class MergeHandler(KerasV3LayerHandler):
         match cls_name:
             case 'Concatenate':
                 rank = len(output_shape)
-                class_name = f'Concatenate{rank}d'
+                class_name = 'Concatenate'
+                op = f'Concatenate{rank}d'
                 config['axis'] = layer.axis
             case 'Dot':
                 msg = (

--- a/hls4ml/model/optimizer/passes/bit_exact.py
+++ b/hls4ml/model/optimizer/passes/bit_exact.py
@@ -221,8 +221,16 @@ def _produce_kif(layer: Layer) -> KIF_t:
 
 @_produce_kif.register
 def _(layer: Input):
-    k = np.ones(get_output_shape(layer), dtype=np.int16)
-    i = f = np.full(get_output_shape(layer), 126, dtype=np.int16)
+    shape = get_output_shape(layer)
+    if layer.attributes.get('trusted', False):
+        precision: FixedPrecisionType = layer.get_output_variable().type.precision
+        k, i, f = precision.signed, precision.integer - precision.signed, precision.fractional
+        k = np.full(shape, k, dtype=np.int16)
+        i = np.full(shape, i, dtype=np.int16)
+        f = np.full(shape, f, dtype=np.int16)
+    else:
+        k = np.ones(shape, dtype=np.int16)
+        i = f = np.full(shape, 126, dtype=np.int16)
     return k, i, f
 
 
@@ -630,8 +638,8 @@ def kif_arrs_to_ints(arr: tuple[np.ndarray, np.ndarray, np.ndarray]):
     return tuple(int(np.max(a)) for a in arr)
 
 
-def produce_kif(layer: Layer) -> KIF_t:
-    if layer.attributes.get('_produce_kif'):
+def produce_kif(layer: Layer, force_reset=False) -> KIF_t:
+    if layer.attributes.get('_produce_kif') and not force_reset:
         return layer.attributes['_produce_kif']
     kif = _produce_kif(layer)
     layer.attributes['_produce_kif'] = kif
@@ -885,7 +893,9 @@ class BitExact(ModelOptimizerPass):
         for node in model.graph.values():
             if node.attributes.get('bit_exact_transformed'):
                 continue
-            produce_kif(node)  # Shrink FixedPointQuantizer bits when possible to be used in backward flow (requested_kif).
+            produce_kif(
+                node, force_reset=True
+            )  # Shrink FixedPointQuantizer bits when possible to be used in backward flow (requested_kif).
 
         for node in model.graph.values():
             if node.attributes.get('bit_exact_transformed'):
@@ -894,12 +904,29 @@ class BitExact(ModelOptimizerPass):
             node.attributes['bit_exact_transformed'] = True
 
         for node in model.graph.values():
-            if node.attributes.get('_produce_kif'):
+            if '_produce_kif' in node.attributes:
                 del node.attributes['_produce_kif']
-            if node.attributes.get('_request_kif'):
+            if '_request_kif' in node.attributes:
                 del node.attributes['_request_kif']
 
         return True
+
+
+def get_output_layers_and_quantizers(
+    node: Layer, layers: list | None = None, quantizers: list | None = None
+) -> tuple[list[Layer], list[FixedPointQuantizer]]:
+
+    layers = layers if layers is not None else []
+    quantizers = quantizers if quantizers is not None else []
+    for _node in get_output_layers(node):
+        if isinstance(_node, FixedPointQuantizer):
+            quantizers.append(_node)
+        elif isinstance(_node, (Reshape, Transpose, Concatenate)):
+            layers.append(_node)
+            get_output_layers_and_quantizers(_node, layers, quantizers)
+        else:
+            raise ValueError(f'Layer {node.name} ({node.class_name}) unexpected input layer chain.')
+    return layers, quantizers
 
 
 class FixInputPrecision(OptimizerPass):
@@ -911,21 +938,17 @@ class FixInputPrecision(OptimizerPass):
         return node.get_output_variable().type.precision.width > 100
 
     def transform(self, model, node: Layer):
-        out_layers: list[FixedPointQuantizer] = get_output_layers(node)  # type: ignore
-        for layer in out_layers:
-            assert isinstance(
-                layer, FixedPointQuantizer
-            ), f'Input {node.name} connected to non-quantizer {layer.name} with non-trivial configuration'
+        layers, out_quantizers = get_output_layers_and_quantizers(node)
 
-        if len(out_layers) == 0:  # Input connected to nothing
+        if len(out_quantizers) == 0:  # Input connected to nothing
             new_type = to_hls4ml_fixed(0, 0, 1, f'{node.name}_t')
             node.get_output_variable().type = new_type
             node.model.config.layer_name_precision[node.name] = str(new_type)
             return False
 
-        sat_modes = [l.SAT for l in out_layers]
+        sat_modes = [l.SAT for l in out_quantizers]
         sat_modes_set = set(sat_modes)
-        rnd_modes = [l.RND for l in out_layers]
+        rnd_modes = [l.RND for l in out_quantizers]
         rnd_modes_set = set(rnd_modes)
         illegal_sat_modes = sat_modes_set - {'WRAP', 'SAT', 'SAT_SYM'}
         illegal_rnd_modes = rnd_modes_set - {'TRN', 'RND'}
@@ -936,7 +959,7 @@ class FixInputPrecision(OptimizerPass):
         if illegal_rnd_modes:
             warn(f'Saturation mode {illegal_rnd_modes} may compromise bit-exactness. Forcing at maximum 24 fractional bits.')
 
-        kifs = [_produce_kif(l) for l in out_layers]
+        kifs = [_produce_kif(l) for l in out_quantizers]
         i = np.max([np.max(i) for _, i, _ in kifs])
         k = np.max([np.max(k) for k, _, _ in kifs])
         if illegal_rnd_modes:
@@ -951,4 +974,15 @@ class FixInputPrecision(OptimizerPass):
             new_type.precision.saturation_mode = 'SAT'
         node.get_output_variable().type = new_type
         node.model.config.layer_name_precision[node.name] = str(new_type)
+        node.attributes['trusted'] = True
+
+        for layer in layers:
+            produce_kif(layer, force_reset=True)
+        for layer in layers:
+            register_precision(layer)
+        for layer in layers:
+            if '_produce_kif' in layer.attributes:
+                del layer.attributes['_produce_kif']
+            if '_request_kif' in layer.attributes:
+                del layer.attributes['_request_kif']
         return False

--- a/hls4ml/model/optimizer/passes/hgq_proxy_model.py
+++ b/hls4ml/model/optimizer/passes/hgq_proxy_model.py
@@ -6,7 +6,7 @@ from warnings import warn
 import numpy as np
 
 from hls4ml.model.attributes import Attribute, TypeAttribute, WeightAttribute
-from hls4ml.model.layers import Activation, Layer, Reshape, register_layer
+from hls4ml.model.layers import Activation, Layer, Reshape, Transpose, register_layer
 from hls4ml.model.optimizer import OptimizerPass, register_pass
 from hls4ml.model.types import FixedPrecisionType, UnspecifiedPrecisionType
 
@@ -100,7 +100,7 @@ class FuseFixedPointQuantizer(OptimizerPass):
         node.attributes['result_t'].precision = precision
         node.attributes['_result_t_propagated'] = True
 
-        if not isinstance(node, Reshape):
+        if not isinstance(node, (Reshape, Transpose)):
             return node
 
         inp_layer = get_input_layers(node)[0]

